# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [2]:
ls

 Volume in drive C has no label.
 Volume Serial Number is 666F-8574

 Directory of C:\Users\Prem\Desktop\SPRINGBOARD_MASTER\6\1588204778_GuidedCapstoneStep4_and_step5HL.ipynb

07/06/2020  10:52 PM    <DIR>          .
07/06/2020  10:52 PM    <DIR>          ..
07/06/2020  10:30 PM    <DIR>          .ipynb_checkpoints
07/05/2020  09:35 PM    <DIR>          __MACOSX
07/06/2020  01:22 AM            34,743 Notebook_stepfour.ipynb
07/06/2020  03:46 AM            40,453 Notebook_stepfourandfive.ipynb
07/06/2020  10:52 PM            48,554 Notebook_stepfourandfive-Copy1.ipynb
07/05/2020  09:28 PM            48,066 step3_output.csv
               4 File(s)        171,816 bytes
               4 Dir(s)  300,634,091,520 bytes free


**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [3]:
file = './step3_output.csv'
df = pd.read_csv(file, index_col=0)
#del df['Unnamed: 0.1']
df.head()

Unnamed: 0.1                 Name    state  summit_elev  vertical_drop  \
0             0       Alyeska Resort   Alaska         3939           2500   
1             1  Eaglecrest Ski Area   Alaska         2600           1540   
2             2     Hilltop Ski Area   Alaska         2090            294   
3             3     Arizona Snowbowl  Arizona        11500           2300   
4             4  Sunrise Park Resort  Arizona        11100           1800   

   base_elev  trams  fastEight  fastSixes  fastQuads  ...  LongestRun_mi  \
0        250      1        0.0          0          2  ...            1.0   
1       1200      0        0.0          0          0  ...            2.0   
2       1796      0        0.0          0          0  ...            1.0   
3       9200      0        0.0          1          0  ...            2.0   
4       9200      0        0.0          0          1  ...            1.2   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0             1610.0           113.0             150.0       60.0   
1              640.0            60.0              45.0       44.0   
2               30.0            30.0             150.0       36.0   
3              777.0           104.0             122.0       81.0   
4              800.0            80.0             115.0       49.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0            669.0          65.0          85.0              150.0   
1            350.0          47.0          53.0               90.0   
2             69.0          30.0          34.0              152.0   
3            260.0          89.0          89.0              122.0   
4            250.0          74.0          78.0              104.0   

   NightSkiing_ac  
0           550.0  
1             0.0  
2            30.0  
3             0.0  
4            80.0  

[5 rows x 27 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [4]:
state = pd.get_dummies(df.state)   

df['state'] = pd.DataFrame(state)

In [5]:
df.head()

Unnamed: 0.1                 Name  state  summit_elev  vertical_drop  \
0             0       Alyeska Resort      1         3939           2500   
1             1  Eaglecrest Ski Area      1         2600           1540   
2             2     Hilltop Ski Area      1         2090            294   
3             3     Arizona Snowbowl      0        11500           2300   
4             4  Sunrise Park Resort      0        11100           1800   

   base_elev  trams  fastEight  fastSixes  fastQuads  ...  LongestRun_mi  \
0        250      1        0.0          0          2  ...            1.0   
1       1200      0        0.0          0          0  ...            2.0   
2       1796      0        0.0          0          0  ...            1.0   
3       9200      0        0.0          1          0  ...            2.0   
4       9200      0        0.0          0          1  ...            1.2   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0             1610.0           113.0             150.0       60.0   
1              640.0            60.0              45.0       44.0   
2               30.0            30.0             150.0       36.0   
3              777.0           104.0             122.0       81.0   
4              800.0            80.0             115.0       49.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0            669.0          65.0          85.0              150.0   
1            350.0          47.0          53.0               90.0   
2             69.0          30.0          34.0              152.0   
3            260.0          89.0          89.0              122.0   
4            250.0          74.0          78.0              104.0   

   NightSkiing_ac  
0           550.0  
1             0.0  
2            30.0  
3             0.0  
4            80.0  

[5 rows x 27 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we don't need in the dataframe for modeling, so it can be dropped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [6]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled = scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [7]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [8]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [9]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [10]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
#explained_variance_score(y_pred)

explained_variance_score(y_test, y_pred)

0.9180291739314158

In [11]:
mean_absolute_error(y_test, y_pred)

5.633604810457884

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [12]:
print(lm.intercept_)

64.04297538878104


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [13]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

Coefficient
Unnamed: 0.1          1.044479
state                 0.017989
summit_elev          11.694564
vertical_drop         3.787694
base_elev             9.310860
trams                 0.268102
fastEight             0.140468
fastSixes             0.213110
fastQuads             0.648777
quad                  1.387485
triple                1.449304
double                0.270228
surface               1.042031
total_chairs          0.675813
Runs                  0.792168
TerrainParks          0.329194
LongestRun_mi         0.132729
SkiableTerrain_ac     0.165589
Snow Making_ac        0.226028
daysOpenLastYear      1.231243
yearsOpen             0.285797
averageSnowfall       1.924904
AdultWeekday         20.126145
projectedDaysOpen     0.311306
NightSkiing_ac        0.245383

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [19]:
F = df.drop(['state'], axis=1)
F.head()

# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
A = F.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
B = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(A)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
A_scaled = scaler.transform(A) 

# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
B = B.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
A_train, A_test, B_train, B_test = train_test_split(A_scaled, B, test_size=0.25, random_state=1)

#all first model set
#from sklearn import linear_model
#from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model2 = lm.fit(A_train,B_train)

# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
B_pred = model2.predict(A_test)

In [20]:
explained_variance_score(B_test, B_pred)

0.918175722940917

In [21]:
mean_absolute_error(B_test, B_pred)

5.630920770197766

In [22]:
print(lm.intercept_)

64.0433144757276


In [23]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), A.columns, columns=['Coefficient'])

Coefficient
Unnamed: 0.1          1.048079
summit_elev          11.675450
vertical_drop         3.782217
base_elev             9.297949
trams                 0.268383
fastEight             0.140678
fastSixes             0.211876
fastQuads             0.648110
quad                  1.387620
triple                1.449593
double                0.269814
surface               1.040591
total_chairs          0.675885
Runs                  0.791166
TerrainParks          0.327462
LongestRun_mi         0.133937
SkiableTerrain_ac     0.165267
Snow Making_ac        0.227022
daysOpenLastYear      1.233707
yearsOpen             0.286107
averageSnowfall       1.926697
AdultWeekday         20.127347
projectedDaysOpen     0.313235
NightSkiing_ac        0.245445

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [24]:
G = df.drop(['state','summit_elev','base_elev'], axis=1)
G.head()

# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
A = G.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
B = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(A)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
A_scaled = scaler.transform(A) 

# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
B = B.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
A_train, A_test, B_train, B_test = train_test_split(A_scaled, B, test_size=0.25, random_state=1)

#all first model set
#from sklearn import linear_model
#from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model2 = lm.fit(A_train,B_train)

# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
B_pred = model2.predict(A_test)

In [25]:
explained_variance_score(B_test, B_pred)

0.9206514527652756

In [26]:
mean_absolute_error(B_test, B_pred)

5.538052121879424

In [27]:
print(lm.intercept_)

64.04558371644728


In [28]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), A.columns, columns=['Coefficient'])

Coefficient
Unnamed: 0.1          1.154595
vertical_drop         0.809968
trams                 0.270001
fastEight             0.023100
fastSixes             0.204847
fastQuads             0.684336
quad                  1.379769
triple                1.428428
double                0.270732
surface               0.993340
total_chairs          0.699173
Runs                  0.903230
TerrainParks          0.250217
LongestRun_mi         0.105421
SkiableTerrain_ac     0.215264
Snow Making_ac        0.153034
daysOpenLastYear      1.186239
yearsOpen             0.296553
averageSnowfall       2.118283
AdultWeekday         20.065517
projectedDaysOpen     0.153854
NightSkiing_ac        0.329625

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.9180291739314158 |5.633604810457884|-|
| Model 2. | 0.9135899636698459| 5.768936822423615|'state'|
| Model 3. | 0.9206514527652756 | 5.538052121879424|'state','summit_elev','base_elev'|

Model Selection:

# Model 3